improve conditional warnings

s-hertel opened the pull request:
##### SUMMARY
Fixes #53428 - Now only non-boolean when statements will trigger this warning
Fixes #56830 - Fix the order for warning on templating delimiters

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/playbook/conditional.py

s-hertel committed:
Fix order for warning on templated conditionals

Fix bare variable warnings when the variable is a boolean

sivel commented on lib/ansible/playbook/conditional.py:
```diff
@@ -114,23 +114,27 @@ def _check_conditional(self, conditional, templar, all_vars):
             # if the conditional is "unsafe", disable lookups
             disable_lookups = hasattr(conditional, '__UNSAFE__')
             conditional = templar.template(conditional, disable_lookups=disable_lookups)
+            if not isinstance(conditional, bool) and bare_vars_warning:
```
I debate whether issuing the warning here is catching some of the edge cases.

Take the following example:

```
- debug:
  when: thing
  vars:
    thing: other
    other: false
```

This is a little contrived for brevity, but is the users intent to evaluate the string `"other"` or have it double interpolated to the value of `False`?

s-hertel replied:
@sivel Good food for thought. I haven't altered what happens for the example task above on this branch versus devel (the var is double interpolated to the value of `False` and a warning is given for the bare var). But if brackets are used instead the bare variable warning doesn't occur.
```
- debug:
  when: thing
  vars:
    thing: "{{ other }}"
    other: false
```
In your example without brackets I'm unsure if the non-bracketed edge case has the wrong behavior. Since there is a bare var in the mix getting a warning for it seems like it could be appropriate. Thoughts?

s-hertel replied:
I feel pretty happy with this as it is. My reasoning is that if the behavior is going to stay for the user when CONDITIONAL_BARE_VARS changes then we don't need to warn them about anything.

s-hertel committed:
changelog

sivel commented:
I left a question on the commit. But looks like I did so in between commits, so it's marked as "outdated" and not shown on the "Files Changed" tab.

s-hertel commented:
@sivel thanks, found it :-)

bcoca commented:
can you add a couple of test cases? one that triggered warning before and does not anymore and one that always triggers warning?

s-hertel committed:
Add tests for cases that should and should not give warnings

If the behavior may change when the default behavior for CONDITIONAL_BARE_VARS becomes False there should be a warning. Boolean type conditionals will not change in behavior so don't warn.

s-hertel committed:
oops, forgot to add files

s-hertel committed:
typo

jamescassell commented:
can we backport this one?

s-hertel commented:
Backport: https://github.com/ansible/ansible/pull/57329
